File: libs/daemon/client/src/mill/client/ServerLauncher.java
package mill.client;

import java.io.*;
import java.net.InetAddress;
import java.net.Socket;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Supplier;

import mill.client.lock.Lock;
import mill.client.lock.Locks;
import mill.constants.DaemonFiles;
import mill.constants.InputPumper;
import mill.constants.ProxyStream;

/// Client side code that interacts with `Server.scala` in order to launch a generic
/// long-lived background daemon.
///
/// The protocol is as follows:
/// - Client:
///   - Take launcherLock
///   - If daemonLock is not yet taken, it means server is not running, so spawn a server
///   - Wait for server socket to be available for connection
/// - Server:
///   - Take daemonLock.
///     - If already taken, it means another server was running
///       (e.g. spawned by a different client) so exit immediately
/// - Server: loop:
///   - Listen for incoming client requests on serverSocket
///   - Execute client request
///   - If launcherLock is released during execution, terminate server (otherwise
///     we have no safe way of terminating the in-process request, so the server
///     may continue running for arbitrarily long with no client attached)
///   - Send `ProxyStream.END` packet and call `clientSocket.close()`
/// - Client:
///   - Wait for `ProxyStream.END` packet or `clientSocket.close()`,
///     indicating server has finished execution and all data has been received
public abstract class ServerLauncher {
  public static class RunWithConnectionResult<A> {
    public final A result;

    public final int exitCode;

    public RunWithConnectionResult(A result, int exitCode) {
      this.result = result;
      this.exitCode = exitCode;
    }
  }

  /// Run a client logic with a connection established to a Mill server (via [#connectToServer]).
  ///
  /// @param connection     the socket connected to the server
  /// @param streams        streams to use for the client logic
  /// @param closeConnectionAfterClientLogic whether to close the connection after running the
  // client logic
  /// @param runClientLogic the client logic to run
  /// @return the exit code that the server sent back
  public static <A> RunWithConnectionResult<A> runWithConnection(
      String debugName,
      Socket connection,
      Streams streams,
      boolean closeConnectionAfterClientLogic,
      Consumer<OutputStream> sendInitData,
      RunClientLogic<A> runClientLogic)
      throws Exception {
    var socketInputStream = new BufferedInputStream(connection.getInputStream());
    var socketOutputStream = new BufferedOutputStream(connection.getOutputStream());
    sendInitData.accept(socketOutputStream);
    socketOutputStream.flush();
    var pumperThread = startStreamPumpers(socketInputStream, socketOutputStream, streams, debugName);
    var result = runClientLogic.run();
    if (closeConnectionAfterClientLogic) socketInputStream.close();
    pumperThread.join();
    return new RunWithConnectionResult<>(result, pumperThread.exitCode());
  }

  /// Run a client logic with a connection established to a Mill server (via [#connectToServer]).
  ///
  /// @param connection     the socket connected to the server
  /// @param closeConnectionAfterClientLogic whether to close the connection after running the
  // client logic
  /// @param runClientLogic the client logic to run
  /// @return the exit code that the server sent back
  public static <A> A runWithConnection(
      String debugName,
      Socket connection,
      boolean closeConnectionAfterClientLogic,
      Consumer<OutputStream> sendInitData,
      RunClientLogicWithStreams<A> runClientLogic)
      throws Exception {
    var socketInputStream = new BufferedInputStream(connection.getInputStream());
    var socketOutputStream = new BufferedOutputStream(connection.getOutputStream());
    sendInitData.accept(socketOutputStream);
    socketOutputStream.flush();
    var result = runClientLogic.run(socketInputStream, socketOutputStream);
    if (closeConnectionAfterClientLogic) socketInputStream.close();
    return result;
  }

  /**
   * Establishes a connection to the Mill server by acquiring necessary locks and potentially
   * starting a new server process if one is not already running.
   *
   * @param daemonDir            the directory where daemon-related files are stored
   * @param serverInitWaitMillis maximum amount of time to wait for the server to start
   * @return a Socket connected to the Mill server
   * @throws Exception if the server fails to start or a connection cannot be established
   */
  public static Socket launchOrConnectToServer(
      Locks locks,
      Path daemonDir,
      String debugName,
      int serverInitWaitMillis,
      InitServer initServer,
      Consumer<ServerLaunchResult.ServerDied> onFailure,
      Consumer<String> log)
      throws Exception {
    var result = ensureServerIsRunning(locks, daemonDir, initServer, log);
    return result.fold(
        success -> {
          try {
            return onServerRunning(daemonDir, debugName, serverInitWaitMillis, log);
          } catch (Exception e) {
            throw new RuntimeException(e);
          }
        },
        alreadyRunning -> {
          try {
            return onServerRunning(daemonDir, debugName, serverInitWaitMillis, log);
          } catch (Exception e) {
            throw new RuntimeException(e);
          }
        },
        processDied -> {
          onFailure.accept(processDied);
          throw new IllegalStateException(processDied.toString());
        });
  }

  /// Invoked when {@link ServerLauncher#ensureServerIsRunning} succeeds.
  public static Socket onServerRunning(
      Path daemonDir, String debugName, int serverInitWaitMillis, Consumer<String> log)
      throws Exception {
    var startTime = System.nanoTime();
    log.accept("Reading server port: " + daemonDir.toAbsolutePath());
    var port = readServerPort(daemonDir, startTime, serverInitWaitMillis);
    log.accept("Read server port, connecting: " + port);
    return connectToServer(
        startTime,
        serverInitWaitMillis,
        port,
        debugName + ". Daemon directory: " + daemonDir.toAbsolutePath());
  }

  public static Integer readServerPort(
      Path daemonDir, long startTimeMonotonicNanos, long serverInitWaitMillis) throws Exception {
    var portFile = daemonDir.resolve(DaemonFiles.socketPort);
    return withTimeout(
        startTimeMonotonicNanos,
        serverInitWaitMillis,
        "Failed to read server port from " + portFile.toAbsolutePath(),
        () -> {
          try {
            return Optional.of(Integer.parseInt(Files.readString(portFile)));
          } catch (IOException e) {
            throw new RuntimeException(e);
          }
        });
  }

  /// Connects to the Mill server at the given port.
  ///
  /// @return a socket that should then be used with [#runWithConnection]
  public static Socket connectToServer(
      long startTimeMonotonicNanos, long serverInitWaitMillis, int port, String errorMessage)
      throws Exception {
    return withTimeout(
        startTimeMonotonicNanos,
        serverInitWaitMillis,
        "Failed to connect to server within " + serverInitWaitMillis + "ms on port " + port + ". "
            + errorMessage,
        () -> {
          try {
            return Optional.of(new Socket(InetAddress.getLoopbackAddress(), port));
          } catch (IOException e) {
            throw new RuntimeException(e);
          }
        });
  }

  public static <A> A withTimeout(
      long startTimeMonotonicNanos,
      long timeoutMillis,
      String errorMessage,
      Supplier<Optional<A>> supplier)
      throws Exception {
    var current = Optional.<A>empty();
    Throwable throwable = null;
    var timeoutNanos = timeoutMillis * 1000 * 1000;
    while (current.isEmpty() && System.nanoTime() - startTimeMonotonicNanos < timeoutNanos) {
      try {
        current = supplier.get();
      } catch (Throwable e) {
        throwable = e;

        //noinspection BusyWait
        Thread.sleep(1);
      }
    }

    if (current.isEmpty()) {
      throw new Exception(errorMessage + " (timeout was " + timeoutMillis + "ms)", throwable);
    }
    return current.get();
  }

  /// Attempts to start a server process using [InitServer] if that is needed.
  ///
  /// After this succeeds you should call [#onServerRunning].
  ///
  /// @param locks      the locks to use for coordination
  /// @param daemonDir  the directory where the server will write its port
  /// @param initServer the function to use to start the server
  public static ServerLaunchResult ensureServerIsRunning(
      Locks locks, Path daemonDir, InitServer initServer, Consumer<String> log) throws Exception {
    Files.createDirectories(daemonDir);

    log.accept("Acquiring the launcher lock: " + locks.launcherLock);
    try (var ignored = locks.launcherLock.lock()) {
      // See if the server is already running.
      log.accept("Checking if the daemon lock is available: " + locks.daemonLock);
      if (locks.daemonLock.probe()) {
        log.accept("The daemon lock is available, starting the server.");
        var launchedServer = initServer.init();

        log.accept("The server has started: " + launchedServer);

        log.accept("Waiting for the server to take the daemon lock: " + locks.daemonLock);
        var maybeLaunchFailed =
            waitUntilDaemonTakesTheLock(locks.daemonLock, daemonDir, launchedServer);
        if (maybeLaunchFailed.isPresent()) {
          var outputs = maybeLaunchFailed.get();
          log.accept("The server " + launchedServer + " failed to start: " + outputs);

          return new ServerLaunchResult.ServerDied(launchedServer, outputs);
        } else {
          log.accept(
              "The server " + launchedServer + " has taken the daemon lock: " + locks.daemonLock);
          return new ServerLaunchResult.Success(launchedServer);
        }
      } else {
        log.accept("The daemon lock is not available, there is already a server running.");
        var startTime = System.nanoTime();
        var pidFile = daemonDir.resolve(DaemonFiles.processId);
        log.accept(
            "Trying to read the process ID of a running daemon from " + pidFile.toAbsolutePath());
        var pid = withTimeout(
            startTime,
            10 * 1000,
            "Could not read the process ID from " + pidFile.toAbsolutePath(),
            () -> {
              try {
                // We need to read the contents of the file and then parse them in a loop because of
                // a race
                // condition where an empty file is created first and only then the process ID is
                // written to it,
                // and thus we can read an empty string from the file otherwise.
                var contents = Files.readString(pidFile);
                return Optional.of(Long.parseLong(contents));
              } catch (IOException | NumberFormatException e) {
                return Optional.empty();
              }
            });
        log.accept("Read PID: " + pid);

        var launchedServer =
            // PID < 0 is only used in tests.
            pid >= 0
                ? new LaunchedServer.OsProcess(ProcessHandle.of(pid)
                    .orElseThrow(
                        () -> new IllegalStateException("No process found for PID " + pid)))
                : new LaunchedServer() {
                  @Override
                  public boolean isAlive() {
                    throw new RuntimeException("not implemented, this should never happen");
                  }

                  @Override
                  public void kill() {
                    throw new RuntimeException("not implemented, this should never happen");
                  }
                };
        return new ServerLaunchResult.AlreadyRunning(launchedServer);
      }
    }
  }

  /// Busy-spins until the server process is running and has taken the `daemonLock`, returning an
  // error if the daemon
  /// process dies.
  private static Optional<ServerLaunchOutputs> waitUntilDaemonTakesTheLock(
      Lock daemonLock, Path daemonDir, LaunchedServer server) throws Exception {
    while (daemonLock.probe()) {
      var maybeLaunchFailed = checkIfLaunchFailed(daemonDir, server);
      if (maybeLaunchFailed.isPresent()) return maybeLaunchFailed;

      //noinspection BusyWait
      Thread.sleep(1);
    }

    return Optional.empty();
  }

  /// Checks if the server process has failed to start.
  private static Optional<ServerLaunchOutputs> checkIfLaunchFailed(
      Path daemonDir, LaunchedServer server) throws IOException {
    if (server.isAlive()) return Optional.empty();

    var outputs = readOutputs(daemonDir);
    return Optional.of(outputs);
  }

  /// Reads the output streams from the server process.
  private static ServerLaunchOutputs readOutputs(Path daemonDir) throws IOException {
    var stdout = daemonDir.toAbsolutePath().resolve(DaemonFiles.stdout);
    var stderr = daemonDir.toAbsolutePath().resolve(DaemonFiles.stderr);

    Optional<String> stdoutStr = Optional.empty();
    if (Files.exists(stdout) && Files.size(stdout) > 0)
      stdoutStr = Optional.of(Files.readString(stdout));

    Optional<String> stderrStr = Optional.empty();
    if (Files.exists(stderr) && Files.size(stderr) > 0)
      stderrStr = Optional.of(Files.readString(stderr));

    return new ServerLaunchOutputs(stdoutStr, stderrStr);
  }

  /**
   * Starts the stream pumpers for the given socket connection to handle input and output streams.
   *
   * @param socketInputStream  the input stream from the server
   * @param socketOutputStream the output stream to the server
   * @return a PumperThread that processes the output/error streams from the server
   */
  static PumperThread startStreamPumpers(
      InputStream socketInputStream, OutputStream socketOutputStream, Streams streams,
      String name
  ) {
    var outPumper = new ProxyStream.Pumper(socketInputStream, streams.stdout, streams.stderr);
    var inPump = new InputPumper(() -> streams.stdin, () -> socketOutputStream, true);
    var outPumperThread = new PumperThread(outPumper, "outPump-" + name);
    outPumperThread.setDaemon(true);
    var inThread = new Thread(inPump, "inPump-" + name);
    inThread.setDaemon(true);
    outPumperThread.start();
    inThread.start();
    return outPumperThread;
  }

  public interface InitServer {
    /// Initializes the server process, returning it.
    LaunchedServer init() throws Exception;
  }

  public interface RunClientLogic<A> {
    /// Runs the client logic.
    A run() throws Exception;
  }

  public interface RunClientLogicWithStreams<A> {
    /// Runs the client logic.
    A run(BufferedInputStream inStream, BufferedOutputStream outStream) throws Exception;
  }

  public static class Result {
    public final int exitCode;
    public final Path daemonDir;

    public Result(int exitCode, Path daemonDir) {
      this.exitCode = exitCode;
      this.daemonDir = daemonDir;
    }
  }

  public static class Streams {
    /// The input stream to send to the server as the stdin.
    public final InputStream stdin;

    /// Server's stdout will be written to this output stream.
    public final OutputStream stdout;

    /// Server's stderr will be written to this output stream.
    public final OutputStream stderr;

    public Streams(InputStream stdin, OutputStream stdout, OutputStream stderr) {
      this.stdin = stdin;
      this.stdout = stdout;
      this.stderr = stderr;
    }
  }

  static class PumperThread extends Thread {
    final ProxyStream.Pumper runnable;

    public PumperThread(ProxyStream.Pumper runnable, String name) {
      super(runnable, name);
      this.runnable = runnable;
    }

    public int exitCode() {
      return runnable.exitCode;
    }
  }
}

File: libs/daemon/client/src/mill/client/ServerLauncher.scala
package mill.client

import mill.client.lock.{Lock, Locks}
import mill.constants.DaemonFiles

import java.io.{BufferedInputStream, BufferedOutputStream, OutputStream}
import java.net.{InetAddress, Socket}

/**
 * Client side code that interacts with `Server.scala` in order to launch a generic
 * long-lived background daemon.
 */
object ServerLauncher {

  /**
   * Configuration that affects whether a daemon needs to be restarted.
   * If any of these values change between client invocations, the daemon should restart.
   */
  case class DaemonConfig(
      millVersion: String,
      javaVersion: String,
      jvmOpts: Seq[String],
      millRepositories: Seq[String]
  ) derives upickle.ReadWriter {

    /**
     * Checks if this config differs from another config.
     * Returns a list of reasons why the daemon should restart, or empty if no restart needed.
     */
    def checkMismatchAgainst(other: DaemonConfig): Seq[String] = {
      val results =
        Option.when(millVersion != other.millVersion) {
          s"Mill version changed ($millVersion -> ${other.millVersion})"
        } ++
          Option.when(javaVersion != other.javaVersion) {
            s"Java version changed ($javaVersion -> ${other.javaVersion})"
          } ++
          Option.when(jvmOpts != other.jvmOpts) {
            s"JVM options changed ($jvmOpts -> ${other.jvmOpts})"
          } ++
          Option.when(millRepositories != other.millRepositories) {
            s"Mill repositories changed ($millRepositories -> ${other.millRepositories})"
          }

      results.toSeq
    }
  }
  object DaemonConfig {

    /** Empty config for cases where config tracking is not needed (e.g., zinc workers) */
    def empty: DaemonConfig = DaemonConfig("", "", Seq.empty, Seq.empty)
  }

  case class Launched(port: Int, socket: Option[Socket], launchedServer: LaunchedServer)
      extends AutoCloseable {
    override def close(): Unit = {
      // Swallow exceptions if the close fails
      try socket.foreach(_.close())
      catch { case _: Exception => }
    }
  }

  /**
   * Run client logic with a connection established to a Mill server.
   */
  def runWithConnection[A](
      connection: Socket,
      closeConnectionAfterClientLogic: Boolean,
      sendInitData: OutputStream => Unit,
      runClientLogic: (BufferedInputStream, BufferedOutputStream) => A
  ): A = {
    val socketInputStream = new BufferedInputStream(connection.getInputStream)
    val socketOutputStream = new BufferedOutputStream(connection.getOutputStream)
    sendInitData(socketOutputStream)
    socketOutputStream.flush()
    val result = runClientLogic(socketInputStream, socketOutputStream)
    if (closeConnectionAfterClientLogic) socketInputStream.close()
    result
  }

  /**
   * Establishes a connection to the Mill server by acquiring necessary locks and potentially
   * starting a new server process if one is not already running.
   *
   * @param config Configuration to check against the running daemon. If any values mismatch,
   *               the old daemon is terminated before starting a new one.
   */

  /**
   * Snapshot the daemon publishes so launchers can show user-visible progress
   * while waiting for an in-flight peer to finish. Mirror of
   * `mill.server.Server.DaemonState`; we duplicate it here so that
   * `libs/daemon/client` doesn't depend on `libs/daemon/server`.
   */
  case class DaemonState(
      pid: Long,
      activeConnections: Int,
      acceptingConnections: Boolean
  ) derives upickle.ReadWriter

  private def readDaemonState(daemonDir: os.Path): Option[DaemonState] = {
    val stateFile = daemonDir / DaemonFiles.daemonState
    try
      if (os.exists(stateFile)) Some(upickle.default.read[DaemonState](os.read(stateFile)))
      else None
    catch { case _: Exception => None }
  }

  /**
   * Block until `daemonLock` is free, meaning the old daemon has fully
   * exited. The daemon-side drain handles waiting for in-flight peer
   * connections to finish before the JVM exits, so this just waits for that
   * shutdown to complete. While we wait, we read `daemonState.json` and emit
   * stderr progress so the user can see why their command is blocked.
   */
  private def waitForOldDaemonToExit(
      daemonLock: mill.client.lock.Lock,
      daemonDir: os.Path
  ): Unit = {
    val pollMillis = 200L
    val progressIntervalMillis = 5_000L
    var lastProgressMillis = 0L
    var announced = false
    while (!daemonLock.probe()) {
      val now = System.currentTimeMillis()
      val shouldEmit = !announced || now - lastProgressMillis >= progressIntervalMillis
      if (shouldEmit) {
        readDaemonState(daemonDir) match {
          case Some(state) if state.activeConnections > 0 =>
            val verb = if (announced) "Still waiting" else "Waiting"
            System.err.println(
              s"[mill] $verb for ${state.activeConnections} command(s) on the running daemon " +
                s"(pid=${state.pid}) to finish before restarting it for a new Mill version..."
            )
            announced = true
            lastProgressMillis = now
          case _ if announced =>
            // Drain finished but the daemon JVM hasn't released `daemonLock`
            // yet (or has wedged in shutdown). Keep the user informed so a
            // long stall isn't a silent hang.
            System.err.println("[mill] Waiting for old daemon to fully shut down...")
            lastProgressMillis = now
          case _ => () // First iteration with no active peers; just poll briefly for JVM exit.
        }
      }
      Thread.sleep(pollMillis)
    }
  }

  def launchOrConnectToServer(
      locks: Locks,
      daemonDir: os.Path,
      serverInitWaitMillis: Int,
      initServer: () => LaunchedServer,
      onFailure: ServerLaunchResult.ServerDied => Nothing,
      log: String => Unit,
      openSocket: Boolean,
      config: DaemonConfig
  ): Launched = {
    log(s"Acquiring the launcher lock: ${locks.launcherLock}")
    val locked = locks.launcherLock.lock()
    try {
      // Check if existing daemon has matching config; if not, request a
      // graceful restart. We hold `launcherLock` across the entire restart so
      // no third launcher can come in mid-handover; peers already attached to
      // the running daemon don't need this lock and run to completion under
      // the daemon's drain (which also writes a `RunCommandResult(101)`
      // response so their launchers retry instead of crashing on EOF).
      val processIdFile = daemonDir / DaemonFiles.processId
      val configFile = daemonDir / DaemonFiles.daemonLaunchFingerprint
      if (os.exists(processIdFile)) {
        val storedOpt: Option[DaemonConfig] =
          if (!os.exists(configFile)) None
          else
            try Some(upickle.default.read[DaemonConfig](os.read(configFile)))
            catch { case _: Exception => None }

        storedOpt.foreach { stored =>
          val mismatchReasons = stored.checkMismatchAgainst(config)
          if (mismatchReasons.nonEmpty) {
            mismatchReasons.foreach(reason => log(reason))
            log(s"Terminating old daemon due to config mismatch: $stored -> $config")
            os.remove(processIdFile, checkExists = false)
            waitForOldDaemonToExit(locks.daemonLock, daemonDir)
            log("Old daemon terminated")
          }
        }
      }

      retryWithTimeout(serverInitWaitMillis, "server launch failed") { () =>
        log("launchOrConnectToServer attempt")

        ensureServerIsRunning(
          locks,
          daemonDir,
          initServer,
          serverInitWaitMillis / 3,
          log,
          config
        ) match {
          case ServerLaunchResult.Success(server) =>
            Some(connectToServer(daemonDir, server, openSocket, log))

          case ServerLaunchResult.AlreadyRunning(server) =>
            Some(connectToServer(daemonDir, server, openSocket, log))

          case processDied: ServerLaunchResult.ServerDied => onFailure(processDied)
        }
      }
    } finally {
      locked.release()
    }
  }

  private def connectToServer(
      daemonDir: os.Path,
      server: LaunchedServer,
      openSocket: Boolean,
      log: String => Unit
  ): Launched = {
    log(s"Reading server port: $daemonDir")
    val port = os.read(daemonDir / DaemonFiles.socketPort).toInt

    log("Read server port")
    val socket = Option.when(openSocket) {
      log(s"Connecting: $port")
      new Socket(InetAddress.getLoopbackAddress, port)
    }

    Launched(port, socket, server)
  }

  def retryWithTimeout[A](
      timeoutMillis: Long,
      errorMessage: String
  )(supplier: () => Option[A]): A = {
    val startTimeMonotonicNanos = System.nanoTime()
    var current: Option[A] = None
    var throwable: Throwable = null
    val timeoutNanos = timeoutMillis * 1000 * 1000

    while (current.isEmpty && System.nanoTime() - startTimeMonotonicNanos < timeoutNanos) {
      try current = supplier()
      catch {
        case e: mill.api.daemon.MillException => throw e
        case e: Throwable =>
          throwable = e
          Thread.sleep(1)
      }
    }

    current.getOrElse(throw new Exception(
      s"$errorMessage (timeout was ${timeoutMillis}ms)",
      throwable
    ))
  }

  /**
   * Attempts to start a server process using initServer if needed.
   */
  def ensureServerIsRunning(
      locks: Locks,
      daemonDir: os.Path,
      initServer: () => LaunchedServer,
      timeoutMillis: Long,
      log: String => Unit,
      config: DaemonConfig
  ): ServerLaunchResult = {
    os.makeDir.all(daemonDir)

    log(s"Checking if the daemon lock is available: ${locks.daemonLock}")
    retryWithTimeout(timeoutMillis, "Failed to determine server status") { () =>
      try {
        if (locks.daemonLock.probe()) {
          log("The daemon lock is available, starting the server.")
          try {
            // Write config file when spawning a new daemon
            os.write.over(
              daemonDir / DaemonFiles.daemonLaunchFingerprint,
              upickle.default.write(config, indent = 2)
            )
            val launchedServer = initServer()

            log(s"The server has started: $launchedServer")

            log(s"Waiting for the server to take the daemon lock: ${locks.daemonLock}")
            waitUntilDaemonTakesTheLock(locks.daemonLock, daemonDir, launchedServer) match {
              case Some(outputs) =>
                log(s"The server $launchedServer failed to start: $outputs")
                Some(ServerLaunchResult.ServerDied(launchedServer, outputs))
              case None =>
                log(s"The server $launchedServer has taken the daemon lock: ${locks.daemonLock}")
                Some(ServerLaunchResult.Success(launchedServer))
            }
          } catch {
            case e: Exception => throw new RuntimeException(e)
          }
        } else {
          log("The daemon lock is not available, there is already a server running.")
          val pidFile = daemonDir / DaemonFiles.processId
          log(s"Trying to read the process ID of a running daemon from $pidFile")
          try {
            val contents = os.read(pidFile)
            val pid = contents.toLong
            log(s"Read PID: $pid")

            val launchedServer =
              if (pid >= 0) {
                LaunchedServer.OsProcess(
                  ProcessHandle.of(pid).orElseThrow(() =>
                    new IllegalStateException(s"No process found for PID $pid")
                  )
                )
              } else LaunchedServer.TestStub // PID < 0 is only used in tests

            Some(ServerLaunchResult.AlreadyRunning(launchedServer))
          } catch {
            case _: java.io.IOException | _: NumberFormatException =>
              log("Read PID exception")
              None
          }
        }
      } catch {
        case _: Exception => None
      }
    }
  }

  /**
   * Busy-spins until the server process is running and has taken the daemonLock.
   */
  private def waitUntilDaemonTakesTheLock(
      daemonLock: Lock,
      daemonDir: os.Path,
      server: LaunchedServer
  ): Option[ServerLaunchOutputs] = {
    while (daemonLock.probe()) {
      if (!server.isAlive) return Some(readOutputs(daemonDir))
      Thread.sleep(1)
    }
    None
  }

  private def readOutputs(daemonDir: os.Path): ServerLaunchOutputs = {
    val stdout = daemonDir / DaemonFiles.stdout
    val stderr = daemonDir / DaemonFiles.stderr

    val stdoutStr = Option.when(os.exists(stdout) && os.size(stdout) > 0)(os.read(stdout))
    val stderrStr = Option.when(os.exists(stderr) && os.size(stderr) > 0)(os.read(stderr))

    ServerLaunchOutputs(stdoutStr, stderrStr)
  }
}
